Add APM documentation (#567)
paketo-buildpacks/paketo-website: content/docs/howto/app-monitor.md

### Example Prerequisites

Since all of the Paketo application monitoring buildpacks support the Java programming languages we'll use the [Java sample applications][samples/java].

To run the examples, clone the repository and change to the Java sample application directory.

{{< code/copyable >}}
git clone https://github.com/paketo-buildpacks/samples
cd samples/java
{{< /code/copyable >}}

Both of the examples assume that the [Paketo Base builder][base builder] is the default builder. If you haven't already, set it as the default now.

{{< code/copyable >}}
pack config default-builder paketobuildpacks/builder:base
{{< /code/copyable >}}

### Example with a Binding

Many of the APM tools require a binding to trigger the tools to be installed and to pass required credentials at runtime. This example uses Azure Application Insights which is one such APM tool.

1. `cd application-insights`.
2. Build an image with the Azure Application Insights Java Agent:

    {{< code/copyable >}}
pack build samples/java --volume "$(pwd)/binding:/platform/bindings/application-insights"
{{< /code/copyable >}}

> 3. To connect to Azure Application Insights at runtime a valid [Instrumentation Key][azure application insights instrumentation key] is required.

Author: ```suggestion
3. To connect to Azure Application Insights at runtime a valid [connection string][azure application insights connection string] is required.
```

[bp/datadog]:https://github.com/paketo-buildpacks/datadog
[bp/dynatrace]:https://github.com/paketo-buildpacks/dynatrace
[bp/elastic-apm]:https://github.com/paketo-buildpacks/elastic-apm
[bp/google-stackdriver]:https://github.com/paketo-buildpacks/google-stackdriver
[bp/jprofiler]:https://github.com/paketo-buildpacks/jprofiler
[bp/new-relic]:https://github.com/paketo-buildpacks/new-relic
[bp/yourkit]:https://github.com/paketo-buildpacks/yourkit

<!-- cnb references -->
[pack]:https://github.com/buildpacks/pack
[platforms]:https://buildpacks.io/docs/concepts/components/platform/

<!-- paketo references -->
[bindings]:{{< ref "/docs/howto/configuration#bindings" >}}
[samples]:https://github.com/paketo-buildpacks/samples
[samples/java]:https://github.com/paketo-buildpacks/samples/tree/main/java

<!-- paketo docs references -->
[base builder]:{{< ref "/docs/concepts/builders#base" >}}

<!-- other references -->
[apm]:https://en.wikipedia.org/wiki/Application_performance_management
[apache-skywalking]:https://skywalking.apache.org/
[appdynamics]:https://www.appdynamics.com/
[aternity]:https://www.aternity.com/application-performance-monitoring/

> [azure application insights instrumentation key]:https://docs.microsoft.com/en-us/azure/azure-monitor/app/create-new-resource#copy-the-instrumentation-key

Author: ```suggestion
[azure application insights connection string]:https://docs.microsoft.com/en-us/azure/azure-monitor/app/sdk-connection-string?tabs=net#finding-my-connection-string
```

Connection credentials will be read from the binding at runtime, with the binding secrets being transformed to system properties with the following the pattern `APPLICATIONINSIGHTS_<KEY>=<VALUE>`.

The buildpack supports the Java and Node.js languages.

## Datadog

The [Datadog][datadog] integration is enabled by setting an environment variable for the [Datadog Buildpack][bp/datadog]. If the environment variable `BP_DATADOG_ENABLED` is set to a truthy value (i.e. true, t, 1 ignoring case) at build-time then an agent appropriate for the application will be contributed to the application image.

Datadog configuration is read from any of the standard Datadog environment variables that are set at runtime. This happens directly, the buildpack does not translate them. You can find the full list of configuration properties in the Datadog documentation [available here](https://docs.datadoghq.com/tracing/setup_overview/setup/java/?tab=containers#configuration).

> The Datadog agent requires a side-car agent to be running in addition to the Java agent. This side-car agent runs outside of the buildpack generated image. The [standard Datadog instructions for your container orchestrator of choice](https://docs.datadoghq.com/tracing/setup_overview/setup/java/?tab=containers#configure-the-datadog-agent-for-apm) can be used to install this agent. The Paketo team also has detailed instructions for [various runtimes available here](https://github.com/paketo-buildpacks/datadog/blob/main/docs/).

Reviewer: ```suggestion
The Datadog agent requires a side-car agent to be running in addition to the Java agent. This side-car agent runs outside of the buildpack generated image. The [standard Datadog instructions for your container orchestrator of choice](https://docs.datadoghq.com/tracing/setup_overview/setup/java/?tab=containers#configure-the-datadog-agent-for-apm) can be used to install this agent. The Paketo team also has detailed instructions for [various runtimes available here](https://github.com/paketo-buildpacks/datadog/blob/main/docs/run-on-docker.md).
```
Should this link to here instead? It says various runtimes but there's only Docker instructions there

Author: That's future-proofing :)  Hopefuly, we'll have instructions for at least Kubernetes as well.

To run the examples, clone the repository and change to the Java sample application directory.

{{< code/copyable >}}
git clone https://github.com/paketo-buildpacks/samples
cd samples/java
{{< /code/copyable >}}

Both of the examples assume that the [Paketo Base builder][base builder] is the default builder. If you haven't already, set it as the default now.

{{< code/copyable >}}
pack config default-builder paketobuildpacks/builder:base
{{< /code/copyable >}}

### Example with a Binding

Many of the APM tools require a binding to trigger the tools to be installed and to pass required credentials at runtime. This example uses Azure Application Insights which is one such APM tool.

1. `cd application-insights`.
2. Build an image with the Azure Application Insights Java Agent:

    {{< code/copyable >}}
pack build samples/java --volume "$(pwd)/binding:/platform/bindings/application-insights"
{{< /code/copyable >}}

3. To connect to Azure Application Insights at runtime a valid [Instrumentation Key][azure application insights instrumentation key] is required.

    {{< code/copyable >}}

> echo "<Instrumentation Key>" > binding/InstrumentationKey

Reviewer: This doesn't seem to work, according to docs it needs to be given a 'Connection String' which is then set to a value of InstrumentationKey=<blah>
```suggestion
echo "InstrumentationKey=<key>" > binding/CONNECTION_STRING
```
It must be capitalised too

Author: OK. This must have changed. I just copied over the previous example. Good catch, thanks.

docker run --rm --tty --publish 8080:8080 \

>   --env SERVICE_BINDING_ROOT=/bindings \

Reviewer: Not a big deal if you follow it properly but can this be shown without overriding the SERVICE_BINDING_ROOT? I feel like bindings are tricky as it is

Author: Yes, if you just use `/platform/bindings` it should work w/out that env variable.

>   --volume "$(pwd)/binding:/bindings/app-insights" \

Reviewer: ```suggestion
  --volume "$(pwd)/binding:/bindings/application-insights" \
```

---
title: "How To Use Application Monitoring With Paketo Buildpacks"
weight: 341
menu:
  main:
    parent: howto
    identifier: monitoring

>     name: "Application Monitoring"

Reviewer 2: ```suggestion
    name: "Enable Application Monitoring"
```
This keeps the verb-phrase style of our other How To menu items.
